### PR TITLE
[1/3] Implementation of a more intelligent Arduino trigger with back channel 

### PR DESCRIPTION
**Pull request by padok**
Related: 

+ https://github.com/JARVIS-MoCap/JARVIS-AcquisitionTool/pull/6
+ https://github.com/JARVIS-MoCap/JARVIS-TriggerFirmware/pull/1
+ https://github.com/JARVIS-MoCap/jarvis-docs/pull/1

WIP - do not merge yet

### CODE REVIEW

File: globals.hpp [hunk 1]
```diff
@@ -91,4 +92,12 @@ void createToolBarButton(QToolButton *button, QAction *action, QIcon icon,
                          bool enabled = true, bool checkable = false,
                          QSize minSize = QSize(20, 20));
 
+// QT connect overload selector
```
Comment (JARVIS-MoCap) [inline globals.hpp:95]: What is this and what does it do? :D

File: src/arduinotrigger.cpp [hunk 2]
```diff
@@ -8,33 +8,134 @@
...
+
+ArduinoTrigger::~ArduinoTrigger() {
+
+    qDebug() << "Trying to stop TriggerCOM";
+    workerThread.requestInterruption();
+    workerThread.quit();
+    workerThread.wait();
+
+    // TODO: call super class destructor
+
```
Comment (JARVIS-MoCap) [inline src/arduinotrigger.cpp:135]: Add 'delete workerClass;' here to make sure the serial port gets closed properly and reconnecting the trigger is possible.

**Review by JARVIS-MoCap — CHANGES_REQUESTED** [review]
This looks great, thank you so much! 

There are two minor issues I found:
1. As you also discovered the serial port is not closed properly when reconnecting a trigger (look at the respective comment for a very easy fix).
2. The first time a recording is started after connecting a trigger it throws a 'Missed Ack' error. It doesn't seem to cause any issues, but having an error show up in the status logs is not ideal.

I'll merge the PR once those two issues are fixed. Thank you so much again for the time and effort you put into this :)

File: globals.hpp (hunk 1 above)
Comment (padok) [reply globals.hpp]: this was indeed redundant to `QOverload<QVariantList>::of` thanks for changing it in 6c3a7377ca325593150528ca77b79524499f8bf3

File: src/arduinotrigger.cpp (hunk 2 above)
Comment (padok) [reply src/arduinotrigger.cpp]: I must have missed that, thanks for adding it in 92e8238a619f040cc6c593e10ae3f6792deb691a

File: src/arduinoserialmessages.hpp [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+// +-------+-------+-------+-------+
+// |         header        | pe_hz |
+// +-------+-------+-------+-------+
+// |          pulse_limit          |
+// +-------+-------+-------+-------+
+// |           delay_us            |
+// +-------+-------+-------+-------+
+struct setup_message_t {
+    msg_header header;
+    uint8_t pulse_hz;     // Frequency in herz 1-255; 0 -> OFF
```
[the remark this answers is not in the document]
Comment (padok) [reply src/arduinoserialmessages.hpp:77]: We can of course increase the size of this variable. I was just unaware that the tool supports cameras that can capture more than 255 fps at this time.

**Pull request merged**